### ISSUE TITLE
[Feature] auto allocation of more than one port for e.g. source games

### ISSUE THREAD

**Issue by AleXburnA** [Ctrlpanel-gg/panel#841]
### Feature Description

Certain pterodactyl eggs needs more than one port to function properly (e.g. ARK needs three: main, rcon and query port).

While its possible to add more than one port allocation to a product, the user still needs to do several steps by hand, after the server is deployed:

- go to the pterodactyl panel and manually add the two additional ports
- and then manually change the variables for rcon and query ports to match with those which got allocated

as this breaks the automated deployment of such eggs, it would be nice to have a solution for that.

### Solution Idea

it would be great if there would be a way to customize egg variables at the product setup. There should be a placeholder for additional ports.

### Additional Info

We had a discussion in Discord about this feature: https://discord.com/channels/787829714483019826/1108138041790431435

**Comment by AleXburnA** [comment]
Yeah I recently saw this feature request. I've also just learned that gameport+1 is optional, so this wouldn't be a big showstopper. 

But the other feature (update specified servers variables, e.g. RCON and Query Port, with the actual allocations) is pretty important IMO as it currently prevents ctrlpanel from being usable for games with more than one port. 

This can be done with the API, I just tried it myself with postman:

1.) POST to **https://{{HOST}}/api/client/servers/{{SERVER_ID}}/network/allocations** to assign the allocations to the server.
2.) PATCH to **https://{{HOST}}/api/application/servers/{{SERVER_APPID}}/startup** to update the servers variables with the learned port numbers.

Maybe this has to be done after the server has finished installing. I am not too sure how hard it is to implement a "watcher" in ctrlpanel which waits until the server has finished installing before it does the reconfiguration.

**Comment by Ferks-FK** [comment]
I think #1090 closes this problem, doesn't it?

**Comment by MrWeez** [comment]
There are a lot of difficulties with adding this feature.
User needs to manually create allocations and assign the appropriate ports in the server startup tab via variables.

**Comment by simbabimba-dev** [comment]
i THINK we can try to implement these using queue workers for the watcher part
about the ports, i think we can recursively find 4 ports that arent free using the this
```
$server->allocation = $this->pterodactyl->getServerAttributes($server->pterodactyl_id)['allocation'];
```
and use the pterodactyl application api endpoint given in the picture to update the variables, 

<img width="881" height="604" alt="Image" src="https://github.com/user-attachments/assets/f94d3aa6-9fae-4085-906e-4cc78eba31aa" />
after that we can just assign them and display this on the server settings page so that the user doesnt have to find them 



### If a game/application absolutely requires a port within a specific range or [original port] + x  type of port
we really cant do anything unless there was a config for this which would add complexity to the server creation as well as product creation process

**Comment by simbabimba-dev** [comment]
Keep in mind this would require the admin/owner of the panel to configure the settings for each egg specially as to what variables ctrlpanel can change and understand them as port variables, otherwise the panel would keep on inserting ports into random variables that can break the functionality of the application

**Comment by MrWeez** [comment]
> i THINK we can try to implement these using queue workers for the watcher part about the ports, i think we can recursively find 4 ports that arent free using the this
> 
> ```
> $server->allocation = $this->pterodactyl->getServerAttributes($server->pterodactyl_id)['allocation'];
> ```
> 
> and use the pterodactyl application api endpoint given in the picture to update the variables,
> 
> <img alt="Image" width="881" height="604" src="https://private-user-images.githubusercontent.com/79574809/541205154-f94d3aa6-9fae-4085-906e-4cc78eba31aa.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njk1MzkwODEsIm5iZiI6MTc2OTUzODc4MSwicGF0aCI6Ii83OTU3NDgwOS81NDEyMDUxNTQtZjk0ZDNhYTYtOWZhZS00MDg1LTkwNmUtNGNjNzhlYmEzMWFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTI3VDE4MzMwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY2MTY5OGIyNmU0YTBjNTMyZGFkNWNmZjg1Y2UwZDNhOWIyOGQ0ZmVlMTA1ZjhlM2YwNzk3NzY2NmI2ZGU3NDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XtfSbJbRxGO83i5D7zwo1T0-Q_LFKAoMDHjr9nqaLdU"> after that we can just assign them and display this on the server settings page so that the user doesnt have to find them
> ### If a game/application absolutely requires a port within a specific range or [original port] + x type of port
> we really cant do anything unless there was a config for this which would add complexity to the server creation as well as product creation process

Yes, it can be implemented, but a lot of actions are needed. Assign allocations during creation, get the port numbers, insert the port numbers into the corresponding variables. This is a fairly extensive task, much more difficult than "just specifying ports". The current codebase already consists of many heavy parts.

**Comment by simbabimba-dev** [comment]
> > i THINK we can try to implement these using queue workers for the watcher part about the ports, i think we can recursively find 4 ports that arent free using the this
> > ```
> > $server->allocation = $this->pterodactyl->getServerAttributes($server->pterodactyl_id)['allocation'];
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and use the pterodactyl application api endpoint given in the picture to update the variables,
> > <img alt="Image" width="881" height="604" src="https://private-user-images.githubusercontent.com/79574809/541205154-f94d3aa6-9fae-4085-906e-4cc78eba31aa.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njk1MzkwODEsIm5iZiI6MTc2OTUzODc4MSwicGF0aCI6Ii83OTU3NDgwOS81NDEyMDUxNTQtZjk0ZDNhYTYtOWZhZS00MDg1LTkwNmUtNGNjNzhlYmEzMWFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTI3VDE4MzMwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY2MTY5OGIyNmU0YTBjNTMyZGFkNWNmZjg1Y2UwZDNhOWIyOGQ0ZmVlMTA1ZjhlM2YwNzk3NzY2NmI2ZGU3NDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XtfSbJbRxGO83i5D7zwo1T0-Q_LFKAoMDHjr9nqaLdU"> after that we can just assign them and display this on the server settings page so that the user doesnt have to find them
> > ### If a game/application absolutely requires a port within a specific range or [original port] + x type of port
> > we really cant do anything unless there was a config for this which would add complexity to the server creation as well as product creation process
> 
> Yes, it can be implemented, but a lot of actions are needed. Assign allocations during creation, get the port numbers, insert the port numbers into the corresponding variables. This is a fairly extensive task, much more difficult than "just specifying ports". The current codebase already consists of many heavy parts.

what if we get like 2 or more available ports and the variables and ask the user to enter the ports?
and the ports entered can only be from the list that we give them

ehhh too much work ig
